README.md
The Nordea Gem
==============

[![Build Status](https://travis-ci.org/matiaskorhonen/nordea.png?branch=master)](https://travis-ci.org/matiaskorhonen/nordea) [![Code Climate](https://codeclimate.com/github/k33l0r/nordea.png)](https://codeclimate.com/github/k33l0r/nordea) [![Gem Version](https://badge.fury.io/rb/nordea.png)](http://badge.fury.io/rb/nordea)

Exchange rates from Nordea Bank.

Installation
------------

Just like any other gem:

    gem install nordea


Usage
-----

### With the Money gem

The Money gem integration works in a similar way to the [eu\_central\_bank](https://github.com/RubyMoney/eu_central_bank) library, mostly because I referenced that gem while building this one. A couple of the methods in the Nordea::Bank class are pretty much identical to those in the eu\_central\_bank gem.

    nordea_bank = Nordea::Bank.new
    Money.default_bank =  nordea_bank

    # Load rates
    nordea_bank.update_rates

    # Exchange 100 EUR to USD
    nordea_bank.exchange(100, "EUR", "USD")

    # Exchange 100 USD to South African rands
    Money.us_dollar(100).exchange_to("ZAR")

    # Exchange 100 Canadian dollars to US dollars
    nordea_bank.exchange_with(Money.new(100, "CAD"), "USD")

    # Update the rates to latest
    nordea_bank.update_rates

    # Save rates to cache file
    cache = '/tmp/nordea_cache.yml'
    nordea_bank.save_rates(cache)

    # Load rates from cache file
    nordea_bank.update_rates(cache)

### Without the Money gem

    exchange_rates = Nordea::ExchangeRates.new

    exchange_rates.currencies
      #=> returns a hash of currencies

    exchange_rates.currencies(true)
      #=> returns a hash of currencies, but forces a data update

    exchange_rates.headers
      #=> returns a hash of the data headers from Nordea

For more information, read the [documentation](http://rubydoc.info/gems/nordea).


About the data and data source
------------------------------

Nordea quotes exchange rates on national banking days at least three times a day:

* in the morning at 8.00,
* at noon at 12.15 and
* in the afternoon at 16.15 (approximate times).

### Useful links:

* [More information about Nordea exchange rates](http://j.mp/Nordea_exchange_rates)
* [More information about the data format](http://j.mp/Rates_for_electronic_processing) [PDF]

### Disclaimer

* This is **not** an *official* gem from Nordea.
* I have **no affiliation** to any entity connected to Nordea Bank.
* I **do not** have any control over the exchange rate data provided, and you use
this gem entirely at **your own risk**.

License
-------

MIT License. Copyright 2011-2013 Matias Korhonen.

See the LICENSE file for more details.


CHANGELOG
2014-11-17 - Alexander Simonov
  * Version 2.1
	* Code refactoring
	* Add support for load rates from cache
	* Remove support for Money 5.x
2014-03-28 - Matias Korhonen
  * Version 2.0.2
  * Fixes the Nordea data URL
  * Ensures compatibility with Money 6.0+

2013-09-10 - Matias Korhonen
  * Version 2.0.1
  * Relaxes the dependency version requirements

2013-09-10 - Matias Korhonen
  * Version 2.0.0
  * Compatible with Money 5.1.1 and above only

2011-06-19 - Matias Korhonen
  * Version 1.1.0
  * Fixes time parsing and thus adds support for Ruby 1.8.7, etc.

2011-06-18 - Matias Korhonen
  * Version 1.0.0
  * Initial release


spec/spec_helper.rb
require 'simplecov'
require 'coveralls'

SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter[
  SimpleCov::Formatter::HTMLFormatter,
  Coveralls::SimpleCov::Formatter
]
SimpleCov.start do
  add_filter 'spec'
  minimum_coverage(76)
end
require 'webmock'
require 'webmock/rspec'
require 'nordea'

# Load support files
Dir["#{File.dirname(__FILE__)}/support/**/*.rb"].each { |f| require f }


spec/bank/bank_spec.rb
require "spec_helper"

describe "Nordea::Bank" do
  let(:bank) do
    Nordea::Bank.new
  end
  
  before(:each) do
    stub_request(:get, "http://openpages.nordea.com/fi/lists/currency/elelctronicExchangeFI.dat").
             to_return(:status => 200, :body => SampleData.raw)
    Money.default_bank = bank
  end

  describe "#exchange" do
    it "returns the correct exchange rates" do
      bank.known_currencies.each do |currency|
        bank.exchange(100, "EUR", currency).cents.should == (SampleData.get_rate(currency) * 100).round
      end
    end
  end

  describe "#exchange_with" do
    it "returns the correct exchange rates" do
      bank.known_currencies.each do |currency|
        bank.exchange_with(Money.new(100, "EUR"), currency).cents.should == (SampleData.get_rate(currency) * 100).round
      end
    end
  end

  describe "#exchange_rates" do
    it "is an instance of Nordea::ExchangeRates" do
      bank.exchange_rates.should be_an_instance_of(Nordea::ExchangeRates)
    end
  end
end

spec/nordea/bank_spec.rb
require 'spec_helper'

describe Nordea::Bank do
  let(:bank) { described_class.new }

  before :each do
    ::Money.default_bank = bank
  end

  context 'update rates from nordea' do
    before(:each) do
      stub_request(:get, 'http://openpages.nordea.com/fi/lists/currency/elelctronicExchangeFI.dat')
        .to_return(status: 200, body: SampleData.raw)
      bank.update_rates
    end

    describe '#exchange' do
      it 'returns the correct exchange rates' do
        bank.known_currencies.each do |currency|
          expect(bank.exchange(100, 'EUR', currency).cents).to eq((SampleData.get_rate(currency) * 100).round)
        end
      end
    end

    describe '#exchange_with' do
      it 'returns the correct exchange rates' do
        bank.known_currencies.each do |currency|
          expect(bank.exchange_with(Money.new(100, 'EUR'), currency).cents).to eq((SampleData.get_rate(currency) * 100).round)
        end
      end
    end
  end

  context 'load from file' do
    let(:file) { File.expand_path('../../support/cache.yml', __FILE__) }
    before :each do
      bank.update_rates(file)
    end
    describe '#exchange' do
      it 'returns the correct exchange rates' do
        bank.known_currencies.each do |currency|
          expect(bank.exchange(100, 'EUR', currency).cents).to eq((SampleData.get_rate(currency) * 100).round)
        end
      end
    end

    describe '#exchange_with' do
      it 'returns the correct exchange rates' do
        bank.known_currencies.each do |currency|
          expect(bank.exchange_with(Money.new(100, 'EUR'), currency).cents).to eq((SampleData.get_rate(currency) * 100).round)
        end
      end
    end
  end

  describe '#save_rates' do
    it 'should call dump_to_yaml' do
      expect(bank.exchange_rates).to receive(:dump_to_yaml).with('blablabla')
      bank.save_rates('blablabla')
    end
  end

  describe '#exchange_rates' do
    subject { super().exchange_rates }
    it { is_expected.to be_a(Nordea::ExchangeRates) }
  end
end


lib/nordea.rb
require 'nordea/bank'
require 'nordea/exchange_rates'
require 'nordea/version'
require 'tzinfo'

# Ruby interface to the Nordea Bank exchange rate data.
module Nordea
  # The default timezone, Europe/Helsinki
  TZ = ::TZInfo::Timezone.get('Europe/Helsinki')
  extend self

  class << self
    # Parses the datetime format used in the Nordea data.
    #
    # @example
    #   Nordea.parse_time("20150101120056") #=> 2015-01-01 10:00:05 UTC
    #
    # @param [String] the datetime string (YYYYMMDDHHmmSS)
    # @return [Time] the string converted into a Time object
    def parse_time(datetime)
      time = Time.utc(datetime[0..3].to_i,
                      datetime[4..5].to_i,
                      datetime[6..7].to_i,
                      datetime[8..9].to_i,
                      datetime[10..11].to_i,
                      datetime[11..12].to_i)

      # Convert the local time to UTC time.
      TZ.local_to_utc(time)
    rescue
      nil
    end

    # Parses the date format used in the Nordea data.
    #
    # @param [String] the date string (YYYYMMDD)
    # @return [Date] the string converted into a Date object
    def parse_date(date)
      ::Date.new(date[0..3].to_i,
                 date[4..5].to_i,
                 date[6..7].to_i)
    rescue
      nil
    end
  end
end


lib/nordea/bank.rb
require 'money'
require 'nordea/exchange_rates'
require 'forwardable'

module Nordea
  # Bank implementation for use with the Money gem.
  #
  # @example
  #   nordea_bank = Nordea::Bank.new
  #   Money.default_bank =  nordea_bank
  #   nordea_bank.exchange(100, "EUR", "USD")
  #   Money.us_dollar(100).exchange_to("ZAR")
  #   nordea_bank.exchange_with(Money.new(100, "CAD"), "USD")
  class Bank < Money::Bank::VariableExchange
    extend ::Forwardable

    def_delegators :exchange_rates, :currencies, :headers, :load_from_yaml, :dump_to_yaml, :force_update

    # Get updated rates from the Nordea server
    #
    # @param [Boolean] force Force or not an update of the exchange rates from Nordea bank
    # @return [Hash] rates available
    def update_rates(file = nil)
      hash_data = if file && ::File.exist?(file)
                    load_from_yaml(file)
                  else
                    force_update
                    currencies
                  end
      @mutex.synchronize do
        hash_data.each_pair do |currency, data|
          rate = data[:middle_rate_for_commercial_transactions]
          set_rate('EUR', currency, rate, without_mutex: true) if known_currencies.include?(currency)
        end
      end

      rates
    end

    # Save rates to the cache file
    def save_rates(file)
      dump_to_yaml(file)
    end

    # Exchange from one currency to another
    #
    # @example
    #   nordea_bank = Nordea::Bank.new
    #   nordea_bank.update_rates
    #   nordea_bank.exchange(100, "EUR", "USD")
    #
    # @param [Integer] cents the amount for the conversion in cents, or equivalent
    # @param [String] from_currency the source currency
    # @param [String] to_currency the target currency
    # @return [Money] the result of the conversion
    def exchange(cents, from_currency = 'EUR', to_currency = 'EUR')
      fail 'Load rates first' unless @rates.size > 0
      exchange_with(Money.new(cents, from_currency), to_currency)
    end

    # Exchanges the given +Money+ object to a new +Money+ object in
    # +to_currency+.
    #
    # @example
    #   nordea_bank = Nordea::Bank.new
    #   nordea_bank.exchange_with(Money.new(100, "CAD"), "USD")
    #
    # @param [Money] from the Money object from which to convert
    # @param [String] to_currency the ISO code for the target currency
    # @return [Money] the new Money object in the target currency
    def exchange_with(from, to_currency)
      rate = get_rate(from.currency, to_currency)
      unless rate
        from_base_rate = get_rate('EUR', from.currency)
        to_base_rate   = get_rate('EUR', to_currency)
        rate           = to_base_rate / from_base_rate
      end
      ::Money.new((from.cents * rate).round, to_currency)
    end

    # Initialize or use an existing Nordea::ExchangeRates object
    #
    # @return [ExchangeRates] an instance of Nordea::ExchangeRates
    def exchange_rates
      @exchange_rates ||= Nordea::ExchangeRates.new
    end

    # List all currencies known to the money gem
    #
    # @return [Array<String>] ISO currency codes
    def money_currencies
      ::Money::Currency.table.keys.map { |c| c.to_s.upcase }
    end

    # List currencies found in the Money gem *and* the Nordea currency data
    #
    # @return [Array<String>]  ISO currency codes
    def known_currencies
      @known_currencies ||= money_currencies & currencies.keys
    end
  end
end


lib/nordea/exchange_rates.rb
# encoding: UTF-8
require 'net/http'
require 'uri'
require 'time'
require 'date'
require 'yaml'

module Nordea
  # Fetch and update Excahnge rates from Nordea Bank.
  #
  # Parses the custom dataformat used by Nordea into more useful Ruby types and
  # objects.
  class ExchangeRates
    # URI to the machine readable data file.
    DATA_URI = URI::HTTP.build(
      host: 'openpages.nordea.com',
      path: '/fi/lists/currency/elelctronicExchangeFI.dat'
    )

    DATA_FILE_KEYS = [
      :file_id, :record_id, :quotation_time, :rate_type,
      :currency_iso_code, :counter_currency_iso_code,
      :middle_rate_for_commercial_transactions,
      :buying_rate_for_commercial_transactions,
      :selling_rate_for_commercial_transactions,
      :buying_rate_for_cash, :selling_rate_for_cash,
      :direction_of_change, :currency_convertability,
      :euro_area, :euro_adoption_date, :currency_expiry,
      :reserved
    ]

    # Hash keys for key-value pairs that should be converted into floats.
    FLOAT_KEYS     = [
      :middle_rate_for_commercial_transactions,
      :buying_rate_for_commercial_transactions,
      :selling_rate_for_commercial_transactions,
      :buying_rate_for_cash,
      :selling_rate_for_cash
    ]

    def initialize
      @currencies = {}
      @headers    = {}
    end

    def currencies
      force_update unless @currencies.size > 0
      @currencies
    end

    def headers
      force_update unless @headers.size > 0
      @headers
    end

    # Load rates from file
    # @param [String] Path to YAML file with sames rates
    #
    def load_from_yaml(yaml_file)
      data        = ::YAML.load_file(yaml_file)
      @headers    = data[:headers]
      @currencies = data[:currencies]
    end

    # Dump rates o file
    # @param [String] Path to YAML file where need to save rates
    def dump_to_yaml(yaml_file)
      data = {
        currencies: currencies,
        headers:    headers
      }
      ::File.open(yaml_file, 'wb') { |file| file.write data.to_yaml }
    end

    private

    def force_update
      lines = fetch_data.lines.to_a
      load_headers(lines.shift)
      @currencies = load_currencies(lines)
    end

    # The batch headers
    #
    # @return [Hash] key-value pairs
    def load_headers(header_line)
      unpacked = header_line.unpack('A4A3A14A120A9')

      @headers = {
        file_id:       unpacked[0], # File ID   = VK01
        record_id:     unpacked[1], # Record ID = 000
        change_time:   unpacked[2], # Exchange rate change time
        notifications: unpacked[3], # Notifications
        reserved:      unpacked[4] # Reserved
      }
      asciify(@headers)

      @headers[:change_time] = ::Nordea.parse_time(@headers[:change_time])
      @headers[:notifications].strip! if @headers[:notifications].respond_to? :strip!
      @headers[:reserved].strip! if @headers[:reserved].respond_to? :strip!
    end

    # List all currencies as a hash. Uses the currencies' ISO codes as keys.
    #
    # @example
    #   e = Nordea::ExchangeRates.new #=> #<Nordea::ExchangeRates:0x00000102102888>
    #   e.currencies["USD"] #=> {
    #   #                                      :file_id => "VK01",
    #   #                                    :record_id => "001",
    #   #                               :quotation_time => 2011-06-17 15:50:01 +0300,
    #   #                                    :rate_type => :list,
    #   #                            :currency_iso_code => "USD",
    #   #                    :counter_currency_iso_code => "EUR",
    #   #      :middle_rate_for_commercial_transactions => 1.427,
    #   #      :buying_rate_for_commercial_transactions => 1.442,
    #   #     :selling_rate_for_commercial_transactions => 1.412,
    #   #                         :buying_rate_for_cash => 1.459,
    #   #                        :selling_rate_for_cash => 1.395,
    #   #                          :direction_of_change => "-",
    #   #                      :currency_convertability => true,
    #   #                                    :euro_area => false,
    #   #                           :euro_adoption_date => nil,
    #   #                              :currency_expiry => false,
    #   #                                     :reserved => ""
    #   # }
    #
    # @param [String] currency_lines force update
    # @return [Hash] a hash off all list currency rates and their properties
    def load_currencies(currency_lines)
      records_array(currency_lines).each_with_object({}) do |record, hash|
        hash[record[:currency_iso_code]] = record
      end
    end

    # Fetches the latest machine readable currency data from the Nordea server.
    #
    # Nordea quotes exchange rates on national banking days at least three times a day:
    # * in the morning at 8.00,
    # * at noon at 12.15 and
    # * in the afternoon at 16.15 (approximate times).
    #
    # @see http://j.mp/Nordea_exchange_rates More information about Nordea exchange rates
    # @see http://j.mp/Rates_for_electronic_processing More information about the data format
    #
    # @return [String] the raw data string.
    def fetch_data
      res = Net::HTTP.start(DATA_URI.host, DATA_URI.port) do |http|
        http.get(DATA_URI.path, 'USER_AGENT' => 'Nordea::Bank gem')
      end

      if res.code =~ /2\d\d/
        res.body
      else
        fail ServerError, 'The server did not respond in the manner in which we are accustomed to.'
      end
    end

    # The data entries as an array of hashes with minimal processing.
    #
    # @return [Array<Hash>] Array of exchange rate records
    def records_array(lines)
      lines.map { |line| collect_line(line) }.select { |rate| rate[:rate_type] == :list }
    end

    def collect_line(line)
      unpacked = line.unpack('A4A3A14A4A3A3A13A13A13A13A13AAAAx7AA42')
      hash     = {}
      DATA_FILE_KEYS.each_with_index do |key, index|
        hash[key] = unpacked[index]
      end
      line_postprocess(hash)
    end

    # Converts the string values from the Nordea data into Ruby objects where
    # possible.
    #
    # @param [Hash] line a line from the data as a Hash
    # @return [Hash] the same line with some values converted into the
    # expected formats
    def line_postprocess(line)
      # Forces the values to be ASCII strings
      asciify(line)

      line[:quotation_time] = ::Nordea.parse_time(line[:quotation_time])
      line[:reserved].strip! if line[:reserved].respond_to? :strip!
      fill_proper_rate_type(line)

      line[:euro_area] = (line[:euro_area] == '1')

      line[:euro_adoption_date] = if line[:euro_area]
                                    Nordea.parse_date(line[:euro_adoption_date])
                                  else
                                    nil
                                  end

      line[:currency_expiry]         = (line[:currency_expiry] == 'K')
      line[:currency_convertability] = (line[:currency_convertability] == 'K')
      floatify(line)

      line
    end

    def fill_proper_rate_type(line)
      line[:rate_type] = if line[:rate_type] == '0001'
                           :list
                         elsif line[:rate_type] == '0002'
                           :valuation
                         end
    end

    def floatify(line)
      FLOAT_KEYS.each do |key|
        str       = line[key]
        float     = "#{str[0, 6]}.#{str[6, 7]}".to_f
        line[key] = float
      end
    end

    def asciify(hash)
      hash.each do |key, value|
        hash[key] = value.force_encoding('ascii') if value.respond_to?(:force_encoding)
      end
    end
  end

  # The exception that's raised if the server doesn't respond correctly.
  class ServerError < StandardError
  end
end
